mrmustard/physics/representations.py

```diff
@@ -0,0 +1,305 @@
+        new._contract_idxs = self._contract_idxs
+        return new
+
+    def __getitem__(self, idx: int | tuple[int, ...]) -> Bargmann:
+        r"""Returns a copy of self with the given indices marked for contraction."""
+        idx = (idx,) if isinstance(idx, int) else idx
+        for i in idx:
+            if i >= self.ansatz.dim:
+                raise IndexError(
+                    f"Index {i} out of bounds for ansatz {self.ansatz.__class__.__qualname__} of dimension {self.ansatz.dim}."
+                )
+        new = self.__class__(self.A, self.b, self.c)
+        new._contract_idxs = idx
+        return new
+
+    def __matmul__(self, other: Bargmann) -> Bargmann:
+        r"""Implements the inner product of ansatze across the marked indices."""
+        if self.ansatz.degree > 0 or other.ansatz.degree > 0:
+            raise NotImplementedError(
+                "Inner product of ansatze is only supported for ansatze with polynomial of degree 0."
+            )
+        Abc = []
+        for A1, b1, c1 in zip(self.A, self.b, self.c):
+            for A2, b2, c2 in zip(other.A, other.b, other.c):
+                Abc.append(
+                    bargmann.contract_two_Abc(
+                        (A1, b1, c1),
+                        (A2, b2, c2),
+                        self._contract_idxs,
+                        other._contract_idxs,
+                    )
+                )
+        A, b, c = zip(*Abc)
+        return self.__class__(math.astensor(A), math.astensor(b), math.astensor(c))
+
+    def trace(self, idx_z: tuple[int, ...], idx_zconj: tuple[int, ...]) -> Bargmann:
+        r"""Implements the partial trace over the given index pairs.
+
+        Args:
+            idx_z: indices to trace over
+            idx_zconj: indices to trace over
+
+        Returns:
+            Bargmann: the ansatz with the given indices traced over
+        """
+        if self.ansatz.degree > 0:
+            raise NotImplementedError(
+                "Partial trace is only supported for ansatze with polynomial of degree ``0``."
+            )
+        if len(idx_z) != len(idx_zconj):
+            msg = f"The number of indices to trace over must be the same for ``z`` and ``z*`` (got {len(idx_z)} and {len(idx_zconj)})."
+            raise ValueError(msg)
+        A, b, c = [], [], []
+        for Abci in zip(self.A, self.b, self.c):
+            Aij, bij, cij = bargmann.complex_gaussian_integral(Abci, idx_z, idx_zconj, measure=-1.0)
+            A.append(Aij)
+            b.append(bij)
+            c.append(cij)
+        return self.__class__(math.astensor(A), math.astensor(b), math.astensor(c))
+
+    def reorder(self, order: tuple[int, ...] | list[int]) -> Bargmann:
+        r"""Reorders the indices of the A matrix and b vector of an (A,b,c) triple."""
+        A, b, c = bargmann.reorder_abc((self.A, self.b, self.c), order)
+        new = self.__class__(A, b, c)
+        new._contract_idxs = self._contract_idxs
```

Comment:
If we change the order, would this contract_idxs would update accordingly?

Reply:
Good question. Actually, we will never have to call `reorder` on a `Bargmann` object when `_contract_idxs` is not an empty tuple, so I could even just return `new`.